enigma-principal/app/src/esgx/epoch_keeper_u.rs

```diff
@@ -21,10 +24,16 @@ extern "C" {
 /// let worker_params: InputWorkerParams = InputWorkerParams { block_number, workers: result.0, stakes: result.1 };
 /// let sig = set_worker_params(enclave.geteid(), worker_params).unwrap();
 /// ```
-pub fn set_worker_params(eid: sgx_enclave_id_t, worker_params: &InputWorkerParams) -> Result<EpochState, Error> {
+#[logfn(DEBUG)]
+pub fn set_worker_params(eid: sgx_enclave_id_t, worker_params: &InputWorkerParams, epoch_state: Option<EpochState>) -> Result<EpochState, Error> {
     let mut retval: EnclaveReturn = EnclaveReturn::Success;
-    let mut nonce_out: [u8; 32] = [0; 32];
-    let mut rand_out: [u8; 32] = [0; 32];
+    println!("Evaluating nonce/seed based on EpochState: {:?}", epoch_state);
+    let (nonce_in, seed_in) = match epoch_state.clone() {
+        Some(e) => (H256::from(e.nonce).0, H256::from(e.seed).0),
+        None => ([0; 32], [0; 32])
+    };
+    println!("Calling enclave set_worker_params with nonce/seed: {:?}/{:?}", nonce_in.to_vec().to_hex(), seed_in.to_vec().to_hex());
+    let (mut nonce_out, mut rand_out) = ([0; 32], [0; 32]);
```

Comment:
Why are you setting these parameters and then override them?

Reply:
Which parameters are being overridden? I'll add more documentation in the code about the nonce and seed. In a nutshell, the purpose of `nonce_in` and `seed_in` is to restore the state of an active epoch. If the optional `EpochState` is given, the enclaves verifies the the inputs match the sealed data. I was trying to limit the number of ecalls but, if you want, I can split it in two: `set_worker_params` and `verify_worker_params` to better separate concerns. I think that it would clarify the intent.